[FSDP] Add `FSDPExtensions` for TP support

awgu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #85738 [FSDP] Add initial `summon_full_params(with_grads=True)`
* #84911 [FSDP] Add `use_orig_params`
* **#85039 [FSDP] Add `FSDPExtensions` for TP support**

This adds `FSDPExtensions` to enable TP + FSDP composability. To be agnostic to both `ShardedTensor` and `DistributedTensor`, the design relies on customizable hooks.

Some notes:
- I preferred the `_ext` prefix (short for "extension") over `_param_extension` simply because it is shorter. It should not matter much because it is purely internal facing.

Commit:
[FSDP] Add `TensorFlattener` for TP support

[ghstack-poisoned]

awgu commented on test/distributed/fsdp/test_fsdp_tp_integration.py:
```diff
@@ -0,0 +1,327 @@
+# Owner(s): ["oncall: distributed"]
```
Either we do not land this PR with this test file included, or I need to clean it up a bit. This is directly taken from the previous TP + FSDP draft PR with some `TensorFlattener` logic added. I used it to verify locally that the `TensorFlattener` implementation at least works for the previous tests.

In my understanding, the tests in this file will not run in normal CI since the test machines have <= 2 GPUs.

Commit:
Update on "[FSDP] Add `TensorFlattener` for TP support"




[ghstack-poisoned]

awgu commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -1299,12 +1299,16 @@ def _materialize_module(
+                and _TORCHDISTX_AVAIL
+                and any(fake.is_fake(p) for p in module.parameters())
+            )
+        except RuntimeError:
```
This is not good. `ShardedTensor` does not support `is_meta`, so we get an error in `__torch_function__()`:
https://github.com/pytorch/pytorch/blob/270e5e519d98868af0166f3a179b286682cfb267/torch/distributed/_shard/sharded_tensor/api.py#L1085-L1087

awgu commented on torch/distributed/fsdp/flat_param.py:
```diff
@@ -311,22 +343,33 @@ def _init_flat_param(
-                    if dtype is None and not param.is_floating_point():
-                        raise ValueError("Integer parameters are unsupported")
+                    if dtype is None:
+                        try:
```
This is not good. `ShardedTensor` does not support `is_floating_point()`, so we get an error in `__torch_function__()`:
https://github.com/pytorch/pytorch/blob/270e5e519d98868af0166f3a179b286682cfb267/torch/distributed/_shard/sharded_tensor/api.py#L1085-L1087

Commit:
Update on "[FSDP] Add `TensorFlattener` for TP support"




[ghstack-poisoned]

Commit:
Update on "[FSDP] Add `TensorFlattener` for TP support"


This PR needs a bit more work:
- Additional hooks needed by TP (beyond the `pre_flatten_transform()` and `post_unflatten_transform()`)
- Fixing `is_meta` and `is_floating_point` in `ShardedTensor` to avoid the `try: ... except: ...`

[ghstack-poisoned]

Commit:
Update on "[FSDP] Add `TensorFlattener` for TP support"


This PR needs a bit more work:
- Additional hooks needed by TP (beyond the `pre_flatten_transform()` and `post_unflatten_transform()`)
- Fixing `is_meta` and `is_floating_point` in `ShardedTensor` to avoid the `try: ... except: ...`

[ghstack-poisoned]

kumpera commented:
This PR has the complete  version with all callbacks we need: https://github.com/pytorch/pytorch/pull/85191

Commit:
Update on "[FSDP] Add `TensorFlattener` for TP support"


This PR needs a bit more work:
- Additional hooks needed by TP (beyond the `pre_flatten_transform()` and `post_unflatten_transform()`)
- Fixing `is_meta` and `is_floating_point` in `ShardedTensor` to avoid the `try: ... except: ...`

[ghstack-poisoned]

Commit:
Update on "[FSDP] Add `TensorFlattener` for TP support"


This PR needs a bit more work:
- Additional hooks needed by TP (beyond the `pre_flatten_transform()` and `post_unflatten_transform()`)

[ghstack-poisoned]

Commit:
Update on "[FSDP] Add `TensorFlattener` for TP support"


This PR needs a bit more work:
- Additional hooks needed by TP (beyond the `pre_flatten_transform()` and `post_unflatten_transform()`)

[ghstack-poisoned]

Commit:
Update on "[FSDP] Add `TensorFlattener` for TP support"


This PR needs a bit more work:
- Additional hooks needed by TP (beyond the `pre_flatten_transform()` and `post_unflatten_transform()`)

[ghstack-poisoned]

Commit:
Update on "[FSDP] Add `TensorFlattener` for TP support"


This PR needs a bit more work:
- Additional hooks needed by TP (beyond the `pre_flatten_transform()` and `post_unflatten_transform()`)

[ghstack-poisoned]

Commit:
Update on "[FSDP] Add `TensorFlattener` for TP support"


This adds `TensorFlattener` to enable TP + FSDP composability. To be agnostic to both `ShardedTensor` and `DistributedTensor`, the design relies on customizable hooks.


[ghstack-poisoned]

Commit:
Update on "[FSDP] Add `TensorFlattener` for TP support"


This adds `TensorFlattener` to enable TP + FSDP composability. To be agnostic to both `ShardedTensor` and `DistributedTensor`, the design relies on customizable hooks.

Some notes:
- I preferred the `_tf` prefix (short for `TensorFlattener`) over `_param_extension` simply because it is shorter. It should not matter much because it is purely internal facing.
- We do not need to include the unit test in the PR when landing. It only tests the customized code paths `_tf_pre_flatten_transform()` and `_tf_post_unflatten_transform()` and not the customized `_tf_chunk_tensor()` and `_tf_pre_load_state_dict_transform()` code paths. If we _do_ want to include it, I may want to clean it up a bit.

[ghstack-poisoned]

Commit:
Update on "[FSDP] Add `TensorFlattener` for TP support"


This adds `TensorFlattener` to enable TP + FSDP composability. To be agnostic to both `ShardedTensor` and `DistributedTensor`, the design relies on customizable hooks.

Some notes:
- I preferred the `_tf` prefix (short for `TensorFlattener`) over `_param_extension` simply because it is shorter. It should not matter much because it is purely internal facing.
- We do not need to include the unit test in the PR when landing. It only tests the customized code paths `_tf_pre_flatten_transform()` and `_tf_post_unflatten_transform()` and not the customized `_tf_chunk_tensor()` and `_tf_pre_load_state_dict_transform()` code paths. If we _do_ want to include it, I may want to clean it up a bit.

[ghstack-poisoned]

Commit:
Update on "[FSDP] Add `TensorFlattener` for TP support"


This adds `TensorFlattener` to enable TP + FSDP composability. To be agnostic to both `ShardedTensor` and `DistributedTensor`, the design relies on customizable hooks.

Some notes:
- I preferred the `_tf` prefix (short for `TensorFlattener`) over `_param_extension` simply because it is shorter. It should not matter much because it is purely internal facing.
- We do not need to include the unit test in the PR when landing. It only tests the customized code paths `_tf_pre_flatten_transform()` and `_tf_post_unflatten_transform()` and not the customized `_tf_chunk_tensor()` and `_tf_pre_load_state_dict_transform()` code paths. If we _do_ want to include it, I may want to clean it up a bit.

[ghstack-poisoned]

Commit:
Update on "[FSDP] Add `TensorFlattener` for TP support"


This adds `TensorFlattener` to enable TP + FSDP composability. To be agnostic to both `ShardedTensor` and `DistributedTensor`, the design relies on customizable hooks.

Some notes:
- I preferred the `_tf` prefix (short for `TensorFlattener`) over `_param_extension` simply because it is shorter. It should not matter much because it is purely internal facing.
- We do not need to include the unit test in the PR when landing. It only tests the customized code paths `_tf_pre_flatten_transform()` and `_tf_post_unflatten_transform()` and not the customized `_tf_chunk_tensor()` and `_tf_pre_load_state_dict_transform()` code paths. If we _do_ want to include it, I may want to clean it up a bit.

[ghstack-poisoned]

Commit:
Update on "[FSDP] Add `TensorFlattener` for TP support"


This adds `TensorFlattener` to enable TP + FSDP composability. To be agnostic to both `ShardedTensor` and `DistributedTensor`, the design relies on customizable hooks.

Some notes:
- I preferred the `_tf` prefix (short for `TensorFlattener`) over `_param_extension` simply because it is shorter. It should not matter much because it is purely internal facing.
- We do not need to include the unit test in the PR when landing. It only tests the customized code paths `_tf_pre_flatten_transform()` and `_tf_post_unflatten_transform()` and not the customized `_tf_chunk_tensor()` and `_tf_pre_load_state_dict_transform()` code paths. If we _do_ want to include it, I may want to clean it up a bit.

[ghstack-poisoned]

fegin commented on torch/distributed/fsdp/_tensor_flattener.py:
```diff
@@ -0,0 +1,107 @@
+from torch.distributed.fsdp._shard_utils import _create_chunk_sharded_tensor
+
+
+class TensorFlattener:
```
This class seems to be an ABC. Shall we make it to ABC?

fegin commented on torch/distributed/fsdp/flat_param.py:
```diff
@@ -968,7 +982,9 @@ def _unflatten(self, as_params: bool) -> None:
         """
         views = self._get_unflat_views(self.flat_param)
-        for view, (param_name, module, _) in zip(views, self.flat_param._param_infos):
+        for i, (view, (param_name, module, _)) in enumerate(
```
nit: enumerate is not needed?

fegin commented on torch/distributed/fsdp/_tensor_flattener.py:
```diff
@@ -0,0 +1,107 @@
+    return tensor
+
+
+def _tf_chunk_tensor(
```
The naming is a little weird to me. We are not flattening a tensor or operating on a flatten tensor but is chunking an unflatten tensor.

awgu replied:
I am renaming from `TensorFlattener` to `FSDPExtensions` and using `_ext` prefix instead of `_tf`.

fegin commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -2527,8 +2527,15 @@ def _sharded_pre_load_state_dict_hook(
             # All-gather the param (ShardedTensor)
-            shards = param.local_shards()
-            local_tensor = cast(torch.Tensor, shards[0].tensor).flatten()
+            param, shards = _tf_pre_load_state_dict_transform(param)
```
Can we let tf_flattner/_tf_pre_load_state_dict_transform to implement this logic? All the error handling logic is either in `_create_chunk_sharded_tensor` or `_gather_state_dict`. It makes sense to enforce tf_flattener/_tf_pre_load_state_dict_transform to implement the same logic.

awgu replied:
Discussed offline: we will keep things as is.

Commit:
Update on "[FSDP] Add `TensorFlattener` for TP support"


This adds `TensorFlattener` to enable TP + FSDP composability. To be agnostic to both `ShardedTensor` and `DistributedTensor`, the design relies on customizable hooks.

Some notes:
- I preferred the `_tf` prefix (short for `TensorFlattener`) over `_param_extension` simply because it is shorter. It should not matter much because it is purely internal facing.
- We do not need to include the unit test in the PR when landing. It only tests the customized code paths `_tf_pre_flatten_transform()` and `_tf_post_unflatten_transform()` and not the customized `_tf_chunk_tensor()` and `_tf_pre_load_state_dict_transform()` code paths. If we _do_ want to include it, I may want to clean it up a bit.

[ghstack-poisoned]

Commit:
Update on "[FSDP] Add `FSDPExtensions` for TP support"


This adds `FSDPExtensions` to enable TP + FSDP composability. To be agnostic to both `ShardedTensor` and `DistributedTensor`, the design relies on customizable hooks.

Some notes:
- I preferred the `_ext` prefix (short for "extension") over `_param_extension` simply because it is shorter. It should not matter much because it is purely internal facing.

[ghstack-poisoned]

awgu commented:
I renamed from `TensorFlattener` to `FSDPExtensions`.
@fegin I changed the class to be an `ABC`.
@kumpera I added your checkpointing unit test.
@rohan-varma @zhaojuanmao The PR is ready for review.

fegin commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -2527,8 +2527,15 @@ def _sharded_pre_load_state_dict_hook(
+                local_tensor = shards[0].flatten()
+            else:
+                # May have no shards, e.g. shape (1, 1, 756)
+                local_tensor = torch.tensor([], dtype=param.dtype, device=self.compute_device)
```
Just realized I already added some checks in https://github.com/pytorch/pytorch/pull/85640.

awgu replied:
I rebased and handled the rebase conflict. Feel free to take a second look to see if it is what you expect.

Commit:
Update on "[FSDP] Add `FSDPExtensions` for TP support"


This adds `FSDPExtensions` to enable TP + FSDP composability. To be agnostic to both `ShardedTensor` and `DistributedTensor`, the design relies on customizable hooks.

Some notes:
- I preferred the `_ext` prefix (short for "extension") over `_param_extension` simply because it is shorter. It should not matter much because it is purely internal facing.

[ghstack-poisoned]

kumpera commented on torch/distributed/fsdp/flat_param.py:
```diff
@@ -152,6 +154,9 @@ class FlatParameter(nn.Parameter):
             ``_numels``, ``_shapes``, and ``_prefixed_param_names``.
         _shared_param_infos (Tuple[SharedParamInfo, ...]): Shared parameter
             info entries; see :class:`SharedParamInfo`.
+        _param_extensions (Tuple[Optional[Any], ...]): Parameter extensions
```
nit: Maybe mention this is experimental and users should not depend on it being around?

Commit:
Update on "[FSDP] Add `FSDPExtensions` for TP support"


This adds `FSDPExtensions` to enable TP + FSDP composability. To be agnostic to both `ShardedTensor` and `DistributedTensor`, the design relies on customizable hooks.

Some notes:
- I preferred the `_ext` prefix (short for "extension") over `_param_extension` simply because it is shorter. It should not matter much because it is purely internal facing.

[ghstack-poisoned]

fegin reviewed (approved):
LGTM

Commit:
Update on "[FSDP] Add `FSDPExtensions` for TP support"


This adds `FSDPExtensions` to enable TP + FSDP composability. To be agnostic to both `ShardedTensor` and `DistributedTensor`, the design relies on customizable hooks.

Some notes:
- I preferred the `_ext` prefix (short for "extension") over `_param_extension` simply because it is shorter. It should not matter much because it is purely internal facing.

[ghstack-poisoned]

awgu commented:
@pytorchbot merge -g
